app/services/messages/operations.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
+
+module Messages
+  # Message operations
+  class Operations < Base
+    attr_accessor :message, :ids, :mentioned_ids
+
+    def initialize(params)
+      @message = params[:message]
+      @ids = []
+      super(params.merge(messageable: @message.messageable))
+    end
+
+    def self.handle_subscriptions(message)
+      new(message).handle_subscriptions
+    end
+
+    def handle_subscriptions
+      format_message_body
+      notify_users unless mentioned_ids.empty?
+      subscribe_users_to_message
+    end
+
+    def subscribe_users_to_message
```

Comment:
I felt that the code in `message_subscription` was easier to read. It's a bit hard to follow lots of mini-functions (my brain stack doesn't go very deep!)

It seems like `notify_users` just adds more users to the list of subscribed users so that could be absorbed into `subscribe_users_to_message` perhaps renamed as `add_direct_mentioned_users`

Reply:
I removed the `notify_users` method. since it just adds subscription, it can be handled in `subscribe_users_to_message`